implementation/src/main/java/io/smallrye/jwt/auth/AbstractBearerTokenExtractor.java

```diff
@@ -0,0 +1,93 @@
+package io.smallrye.jwt.auth;
+
+import java.util.function.Function;
+
+import javax.inject.Inject;
+
+import org.eclipse.microprofile.jwt.JsonWebToken;
+import org.jboss.logging.Logger;
+
+import io.smallrye.jwt.auth.cdi.PrincipalProducer;
+import io.smallrye.jwt.auth.principal.JWTAuthContextInfo;
+import io.smallrye.jwt.auth.principal.JWTCallerPrincipal;
+import io.smallrye.jwt.auth.principal.JWTCallerPrincipalFactory;
+import io.smallrye.jwt.auth.principal.ParseException;
+
+/**
+ * Common functionality for classes extracting Bearer tokens from HTTP request
+ * headers (including Cookie) and converting the token string to a
+ * {@link JsonWebToken}.
+ *
+ *
+ * @author Michael Edgar {@literal <michael@xlate.io>}
+ */
+public abstract class AbstractBearerTokenExtractor {
+
+    private static Logger logger = Logger.getLogger(AbstractBearerTokenExtractor.class);
+
+    @Inject
+    private JWTAuthContextInfo authContextInfo;
+
+    @Inject
+    private PrincipalProducer producer;
```

Comment:
I'm not sure about these 2 injections. Well, it all looks neat, I agree , but I have the practical concerns. 

My proposal was only to reuse the actual extraction of the token. These injections enforce the integrations to extend this class and there could be no space to extend (well, Quarkus/Thornail can extend but the injection is not done directly into Undertow authentication mechanisms, it does not work in Quarkus at least).

This is why I'd rather have private static classes customizing the extraction only.
Lets take it one step at a time :-) I'll need to refactor Throntail to reuse JAX-RS feature/filters and both Thorntail and Quarkus to reuse the token extraction code (from the headers only). At the next stage we can concentrate on reusing the validation code, but without having to extend, may be we can introduce a ValidationProducer which will take the token value only and return JsonWebToken, etc

Thanks

Reply:
@sberyozkin - I think I'm following you. Bottom line is that I can't assume CDI for the abstract token extraction, correct?

Reply:
@sberyozkin - I refactored out the inheritance between the token extraction class and the authorization filter and EE security mechanism. Let me know if that is alignment with your proposed solution.